src/Language/PureScript/Scheme/CodeGen/Scheme.hs

```diff
@@ -15,19 +15,45 @@ app name args = List ((Symbol name):args)
 t :: SExpr
 t = Symbol "#t"
 
+stringEqQ' :: SExpr
+stringEqQ' = Symbol "string=?"
+
+stringHash' :: SExpr
+stringHash' = Symbol "string-hash"
+
 
 -- Scheme special forms --------------------------------------------------------
 
 define :: Text -> SExpr -> SExpr
 define name expr = List [Symbol "define", Symbol name, expr]
 
+quote :: SExpr -> SExpr
+quote x = app "quote" [x]
+
 lambda :: [Text] -> SExpr -> SExpr
 lambda formals expr =
   List [Symbol "lambda", List $ map Symbol formals, expr]
 
 lambda1 :: Text -> SExpr -> SExpr
 lambda1 formal expr = lambda [formal] expr
 
+let_ :: [(Text, SExpr)] -> SExpr -> SExpr
+let_ bodies expr = List [Symbol "let", List $ fmap go bodies, expr]
+  where
+    go :: (Text, SExpr) -> SExpr
+    go (var, expr') = List [Symbol var, expr']
+
+-- (cond (test expr) ... (else expr))
+condWithElse :: [(SExpr, SExpr)] -> SExpr -> SExpr
+condWithElse clauses elseExpr = cond (clauses ++ [(Symbol "else", elseExpr)])
+
+-- (cond (test expr) ...)
+cond :: [(SExpr, SExpr)] -> SExpr
```

Comment:
Is there any value in having a `cond` without an `else` clause?

If our implementation is wrong and we don't match on any of the clauses then this will presumably return `nil`, which sure is going to be _hard_ to debug?

I'd prefer if we'd always emit an `else` clause, where we'd throw an "internal error" when we don't have an `else` clause at the application level. In this way if our implementation is wrong we'd have a pointer at where things blow up.

In other words, I think `cond` should be implemented with `condWithElse`, where we pass an `else` case that throws.

Reply:
I agree with your suggestion to always use `condWithElse` instead of `cond`.

As of now we're caling `cond` (without else clause) only at [`Transpiler.hs:129`](https://github.com/turlando/purescm/blob/1f7f27f70da735134de1804b8ed9802610e1a45b/src/Language/PureScript/Scheme/CodeGen/Transpiler.hs#L129) when handling case expressions.

In this case I think we're safe to have an exception raised if we hit the else branch.

I would still keep `cond` in `Scheme.hs` though since the cond (without else) is effectively possible in Scheme and provides a tiny lower abstraction for `condWithElse`. Maybe we can make it private?

Reply:
> Maybe we can make it private?

I don't see the point in keeping an unsafe function even if private. What I mean by "I think `cond` should be implemented with `condWithElse`" (instead of the other way around as it is now) is this:

```haskell
condWithElse :: [(SExpr, SExpr)] -> Maybe SExpr -> SExpr
condWithElse clauses maybeElseExpr
  = app "cond"
  $ map (\(test, expr) -> List [test, expr])
  $ clauses <> case maybeElseExpr of
      Nothing -> [Symbol "raise", String "Internal error while handling expression etc etc"]
      Just elseExpr -> [Symbol "else", elseExpr]

cond :: [(SExpr, SExpr)] -> SExpr
cond clauses = condWithElse clauses Nothing
```

In this way we could export both `cond` and `condWithElse`, and they would both be safe :slightly_smiling_face: